### ISSUE TITLE
When Character changes walking dir (loop), don't reset animation frame

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#2126]
As per user request...

**Describe the problem**
When Character walks and changes walking direction, it's view loop is changed, and animation frame reset to 0. Some users think that it's not good and would prefer the frame number kept, so that transition looks smoother. I guess this is specifically important for games with smooth animation and diagonal walking loops, where loop may change often.

**Suggested change**
~Add a Character property (in both editor and engine) that tells whether this character should keep frame or reset frame when changing walking loop. If this property is set to "keep", then the frame is retained if possible. I think the rule should be simple: if a new loop has this frame, then use it, if it has less frames, then reset to 0.~

~NOTE: Personally, I'm not sure if this is has to be a property, and not a fixed behavior. But then, maybe in some cases it looks better if animation is reset on turn...
If a property is preferred, then some good clear name has to be invented for it.~

UPDATE: It turned out to be a mistake in WalkStraight function, as opposed to Walk, which keeps previous frame already.
The details are explained in this comment below:
https://github.com/adventuregamestudio/ags/issues/2126#issuecomment-1715318354

**Comment by ivan-mogilko** [comment]
Hm, in such case a switch in character properties is a must.

Actually, for the best effect I'd imagine there has to be transition rule, where a frame A in loop X is synced with frame B in loop Y, but that would require to design a whole new method of configuring this.

**Comment by messengerbag** [comment]
@ericoporto has a good point. In your original proposal, @ivan-mogilko, the behavior would vary "randomly" for the same walkcycle depending on which frame the turn happened to fall on (e.g., if the side views have 12 frames and the up/down views have 6, then if a turn from side view to up occurred on frame 4, it would use frame 4 of the up cycle, which wouldn't match and appear to glitch, but if it occurred on frame 7 it would reset to frame 0 and appear to pause: the effect would be sometimes a pause and sometimes a skip), which seems undesirable.

So it probably only makes sense to match the frame if the views have exactly the same number of frames.

If you do want to support non-matching frame counts, a more complex but universal way would be to calculate it as `newFrameNumber = currentFrameNumber / currentViewFrameCount * newViewFrameCount` (you might need to subtract 1 from each factor to exclude the standing frame), thus keeping the same proportion of progress through the animation cycle.

**Comment by ivan-mogilko** [comment]
So, apparently I misunderstood what user had in mind when asking for this, AND I also used a inconvenient test case.

I thought that the problem is that AGS resets frame while Character is walking along some path, and turning. That would sound like a mistake. But after I did a better test, with longer animations and distinct sprites, it turned to be not the case. AGS keeps the current walking frame when turning, and actually does this even if there's a difference in frame numbers. It only resets the frame to 0 when it exceeds the number of frames in the current loop.

In fact, it keeps the last frame even when **a new walking destination was ordered**.

So, it seems I don't understand the original problem. I will have to investigate the user's case again.

**Comment by ivan-mogilko** [comment]
I found what the problem is. AGS has two methods: Character.Walk and Character.WalkStraight.
Both of them end up calling `walk_character` function eventually, which does the main job:

https://github.com/adventuregamestudio/ags/blob/7a9211b93ead00c2a245217a3dc2b85a9e9b565c/Engine/ac/character.cpp#L1018
https://github.com/adventuregamestudio/ags/blob/7a9211b93ead00c2a245217a3dc2b85a9e9b565c/Engine/ac/character.cpp#L1048

`walk_character` has this logic: it remembers the previous animation frame before resetting an old move and starting a new move, and then reassigns it:

https://github.com/adventuregamestudio/ags/blob/7a9211b93ead00c2a245217a3dc2b85a9e9b565c/Engine/ac/character.cpp#L1714-L1727

But WalkStraight has a separate StopMoving call before anything can be saved:

https://github.com/adventuregamestudio/ags/blob/7a9211b93ead00c2a245217a3dc2b85a9e9b565c/Engine/ac/character.cpp#L1031

---

To summarize, everything is goes through the `walk_character`, but WalkStraight has 1 extra call to StopMoving, which breaks this logic and resets some things before last state may be saved.

I don't know if this was intentional, possibly this is just an oversight.

The whole problem then is that WalkStraight resets things, while it should not. The fix may be then to simply remove a call to StopMoving from WalkStraight, and let  `walk_character` do what it is supposed to instead.

This does not need any new options. And any further adjustments to the frame selection algorithm should be done inside `walk_character`, but as a separate task (I probably won't do anything like that now).

**Comment by ivan-mogilko** [comment]
Fixed by 79d692a068fba8bafa99ad183484d86979d73132.
